README.txt
This is a CSE 491/Web dev project; see http://msu-web-dev.readthedocs.org/.

There are 3 sub-versions of server.py:
1. server.py is the standard that runs app
2. server-qx.py is the WSGI standard test server
3. server-imageapp.py runs the image WSGI app that uses cookies

1. Running server.py on arctic.
===============================

In a subdirectory on arctic.cse.msu.edu, create a Python `virtualenv
<http://www.virtualenv.org/en/latest/>`__ using Python 2.7::

   python2.7 -m virtualenv cse491.env

Now, activate this virtualenv::

   source cse491.env/bin/activate.csh

(From now on, you'll want to make sure to have this virtualenv active
whenever you're working on server.py.)

Next, go to your cse491-serverz directory and run server.py::

   python server.py

Use CTRL-C to quit the server.




ChangeLog
2014-01-07  Titus Brown  <t@Titus-MacBook-Air-2.local>

 * Initialized repository, created server.py and README.

2014-01-14  Tao Feng

 * Modify the code to return an HTTP 1.0 response, 
 * containing '200 OK' response line, 
 * a 'Content-type' of text/html, 
 * and a message body saying '<h1>Hello, world<h1> this is beckhamer’s Web server.'

2014-01-23 Tao Feng

 * Update server.py code to grab the request data (use 'c.recv(1000)'), 
 * extract the 'path' component, and return different HTML content for the following request URLs:
 * /
 * /content
 * /file
 * /image
 * Modify '/' to return HTML that contains links to /content, /file, and /image.
 * Mofify the handle_connection function to handle POST requests separately from GET.
 * Add new file 'test_server.py' and pass all the tests.

2014-01-30 Tao Feng

 * Update server.py code to parse form/GET and form/POST
 * Modify main function to split different pages into different functions.
 
2014-02-06 Tao Feng

 * Modify POST code handling code to properly handle requests that have a Content-Type of multipart/form-data.
 * Modify send-post-request to send multipart/form-data.
 * Fix code to work with arbitrary size requests.
 * HTML and templating.
 * Modify code to return a "404 Not Found" status.
 * Check code coverage.
 
2014-02-13 Tao Feng

 * Move all of content-creation code into 'app.py', and refactor it to look like a WSGI application.
 * Refactor the handle_connection function to run WSGI apps.
 * Template inheritance and proper HTML using jinja2.
 
2014-02-20 Tao Feng

 * In 'server.py', write a function that serves a file with the appropriate content-type (image/jpeg and text/plain).
 * Add file.txt and image.jpg.
 * 'server-Quixote.py' is to test WSGI server works with all three of the Quixote demo apps.
 * Use the wsgiref validator to evaluate the WSGI app.
 
2014-02-27 Tao Feng

 * Add cookie header handling into WSGI server.
 * Merge in 'run-imageapp.py' and the 'imageapp/' directory.
 * Modify the imageapp to display the latest image on the front (index) page, not just the '/image' page.
 
2014-03-20 Tao Feng

 * Implement command line options in 'server.py' to run the following WSGI apps: imageapp, quixote.demo.altdemo and myapp.
 * Add twill-tests files and write a twill test 'myapp-1.twill' to execute all of the URLs and checks the return code (200).

 * Implement Project 4: support JPEG and TIFF format.
 * In 'image.py', add another parameter 'format' in the function 'add_image'.
 * In 'root.py', extract the file format using 'the_file.base_filename.split('.')[1]' and set the proper content type as img[1].
 * It is noted that Chrome and Firefox do not support TIFF format, but IE does. 
 
2014-03-27 Tao Feng

 * Integrate the quotes app into 'server.py'.
 * Integrate the chats app into 'server.py'.
 
 * Implement Project 5: Add a URL that provides a list of images, resized to thumbnails.
 * The URL is 'list.html' and the corresponding function 'image_list' is added in 'root.py'. 

2014-04-03 Tao Feng

 * Implement Project 3: Swipe a template 'grapes.css' from oswd.org and put it in place on imageapp server.

 * Implement Project 12: Implement AJAX image upload with in 'upload.html' with Javascript. 

2014-04-10 Tao Feng
 * Integrate the cookie app into 'server.py'.
 * Change imageapp to store its images in a SQL database.
 
 * Implement Project 6 and 7 using sqlite database so that the 'metadata' of the image can be added and searched.

2014-04-17 Tao Feng
 * Implement Project 22: Make a thumbnail list using server-side resizing using PIL.

 * Implement Project 23: Write a URL handler so that each image is referred to by its number, e.g. /image_raw?id=1
 * If 'id' is out of range, it will redirect to the latest image.

2014-04-24 Tao Feng
 * Implement Project 8: Add the ability to comment on images.

 * Implement Project 24: Make a generally useful fake data set to support #6.

 * 10 projects implemented are 3, 4, 5, 6, 7, 8, 12, 22, 23, 24.


hw8.py
## for parsing path and query string
from urlparse import urlparse, parse_qs

## for fieldstorage
from StringIO import StringIO
import cgi

## for the templates
import jinja2
import os



def simple_app(environ, start_response):
  
    # corrospond requests to .html files
    response = {
		'/'		: 'index.html',		\
		'/content'	: 'content.html',	\
		'/file'		: 'file.html',		\
		'/image'	: 'image.html',		\
		'/form'		: 'form.html',		\
		'/submit'	: 'submit.html',	\
		}
    
    # setup and run jinja2 templates
    loader = jinja2.FileSystemLoader('./templates')
    jinja2_environment = jinja2.Environment(loader=loader)
    
    
    # if any values are stored in the URL, this will get them
    values = parse_qs(environ['QUERY_STRING'])
    # initialize empty dictionary for POST requests that can be updated
    # with the contents of the awful cgi thing
    new_values = {}
    
    # create field storage object
    WSGIinput = cgi.FieldStorage(fp=environ['wsgi.input'], 
		 headers = environ,
		 environ = {'REQUEST_METHOD':'POST'} )
		      
    if environ['REQUEST_METHOD'] is 'POST':
      
	print "WSGIinput.keys()"
	print WSGIinput.keys()
	
	for key in WSGIinput.keys():
	    new_values[key] = WSGIinput[key].value
	    print "%s = %s \r\n" % (key, new_values[key])
	
	values.update(new_values)
	print 'values updated!' 
    
    
    # Try to connect to requested page
    if environ['PATH_INFO'] in response:
	response_status =	'200 OK'
	template = jinja2_environment.get_template(response[environ['PATH_INFO']])
    else:
	values['page'] = environ['PATH_INFO']
	response_status = '404 Not Found'
	template = jinja2_environment.get_template('404.html')
    
    # initialize return data
    data = []
    # set response_headers depending on the page content type
    if environ['PATH_INFO'] == '/image':
      response_headers = [('Content-type', 'image/jpeg')]
      # return image
      data.append(open_file('image.jpg') )
      
    elif environ['PATH_INFO'] == '/file':
      response_headers = [('Content-type', 'text/plain')]
      # return text file
      data.append(open_file('file.txt') )
      
    else:
      response_headers = [('Content-type', 'text/html')]
      # load template depending on the PATH_INFO
      # using jinja2's render to get the return string
      response_html = template.render(values).encode('latin-1', 'replace')
      data.append(response_html)
    
    print 'tried '+ environ['PATH_INFO']
    
    
    # start_response before building from template
    # does the app.send() for the status and headers
    start_response(response_status, response_headers)
    
    
    
    # return a list (actually a generator) titled data
    # which will be sent in the handle connection function
    return data
    
def open_file(file_name):
  '''Takes string of file name, returns open().read() as object'''
  file_obj = open(file_name, "rb")
  data = file_obj.read();
  file_obj.close()
  return data
    
# still needed?
def make_app():
    return simple_app
    
    